simplemodels/models.py

```diff
@@ -98,19 +76,53 @@ class Meta:
         # TODO: it might make sense to add option to raise an error if unknown
         # field is given for the document
 
-    def __init__(self, **kwargs):
-        kwargs = self._unprotect_fields(kwargs)
-        kwargs = self._clean_kwargs(kwargs)
+    def __init__(self, data=None, **kwargs):
+        if data is None:
+            data = {}
+
+        if self._meta['ALLOW_EXTRA_FIELDS']:
+            self._fields = copy.deepcopy(self._fields)
 
-        # dict init
-        prepared_fields = self._prepare_fields(kwargs)
-        super(Document, self).__init__(**prepared_fields)
+        if not isinstance(data, MutableMapping):
+            raise ModelValidationError("Data must be instance of mapping, but got '%s'!" % type(data))
+
+        data = copy.deepcopy(data)
+
+        data = self._clean_data(data)
+
+        self._prepare_fields(data, **kwargs)
         self._post_init_validation()
 
-    def _prepare_fields(self, kwargs):
+    def __getitem__(self, name):
+        return getattr(self, name)
+
+    def __setitem__(self, name, value):
+        setattr(self, name, value)
+
+    def __delitem__(self, name):
+        delattr(self, name)
+
+    def __iter__(self):
+        return iter(
+            {k: v for k, v in self._fields.items() if self.get(k) is not None or not self._meta['OMIT_MISSED_FIELDS']}
```

Comment:
Split it into several lines and may be add docstring. I have to read several times to understand why iterator looks like that